core/src/main/java/org/infinispan/remoting/rpc/RpcManager.java

```diff
@@ -21,186 +22,17 @@
-    * @deprecated this method may be removed in the future. Use {@link #invokeRemotelyInFuture(java.util.Collection, org.infinispan.commands.ReplicableCommand, RpcOptions, org.infinispan.commons.util.concurrent.NotifyingNotifiableFuture)}
-    */
-   @Deprecated
-   void invokeRemotelyInFuture(Collection<Address> recipients, ReplicableCommand rpc, boolean usePriorityQueue, NotifyingNotifiableFuture<Object> future);
-
-   /**
-    * The same as {@link #invokeRemotelyInFuture(java.util.Collection, org.infinispan.commands.ReplicableCommand,
-    * boolean, org.infinispan.commons.util.concurrent.NotifyingNotifiableFuture)} except that you can specify a timeout.
-    *
-    * @param recipients       recipients to invoke remote call on
-    * @param rpc              command to execute remotely
-    * @param usePriorityQueue if true, a priority queue is used
-    * @param future           the future which will be passed back to the user
-    * @param timeout          after which to give up (in millis)
-    * @deprecated this method may be removed in the future. Use {@link #invokeRemotelyInFuture(java.util.Collection, org.infinispan.commands.ReplicableCommand, RpcOptions, org.infinispan.commons.util.concurrent.NotifyingNotifiableFuture)}
-    */
-   @Deprecated
-   void invokeRemotelyInFuture(final Collection<Address> recipients, final ReplicableCommand rpc, final boolean usePriorityQueue, final NotifyingNotifiableFuture<Object> future, final long timeout);
-
-   /**
-    * The same as {@link #invokeRemotelyInFuture(java.util.Collection, org.infinispan.commands.ReplicableCommand,
-    * boolean, org.infinispan.commons.util.concurrent.NotifyingNotifiableFuture, long)} except that you can specify a response mode.
-    *
-    * @param recipients       recipients to invoke remote call on
-    * @param rpc              command to execute remotely
-    * @param usePriorityQueue if true, a priority queue is used
-    * @param future           the future which will be passed back to the user
-    * @param timeout          after which to give up (in millis)
-    * @param ignoreLeavers    if {@code true}, recipients that leave or have already left the cluster are ignored
-    *                         if {@code false}, a {@code SuspectException} is thrown when a leave is detected
-    * @deprecated this method may be removed in the future. Use {@link #invokeRemotelyInFuture(java.util.Collection, org.infinispan.commands.ReplicableCommand, RpcOptions, org.infinispan.commons.util.concurrent.NotifyingNotifiableFuture)}
-    */
-   @Deprecated
-   void invokeRemotelyInFuture(Collection<Address> recipients, ReplicableCommand rpc,
-                               boolean usePriorityQueue, NotifyingNotifiableFuture<Object> future,
-                               long timeout, boolean ignoreLeavers);
+   CompletableFuture<Map<Address, Response>> invokeRemotelyAsync(Collection<Address> recipients,
```

Comment:
@galderz I briefly considered it, but it would require us to create `n` extra `CompletableFuture` instances for each RPC. And it would also make it impossible for us to use JGroups anycast/broadcast, which give us a single `GroupRequest`.

`CompletableFuture.anyOf` wouldn't work for us either, because we need to be able to ignore some responses - like `ResponseMode.WAIT_FOR_VALID_RESPONSE` does. The `CompletableFuture` returned by `anyOf` would return immediately if one of the futures finishes with a `SuspectException`, for instance, and we want to ignore those when we send a `ClusteredGetCommand`.